snuba/migrations/snuba_migrations/transactions/0001_transactions.py

```diff
@@ -0,0 +1,107 @@
+    Column("transaction_name", LowCardinality(String())),
+    Column("transaction_hash", Materialized(UInt(64), "cityHash64(transaction_name)")),
+    Column("transaction_op", LowCardinality(String())),
+    Column("transaction_status", WithDefault(UInt(8), str(UNKNOWN_SPAN_STATUS))),
+    Column("start_ts", DateTime()),
+    Column("start_ms", UInt(16)),
+    Column("finish_ts", DateTime()),
+    Column("finish_ms", UInt(16)),
+    Column("duration", UInt(32)),
+    Column("platform", LowCardinality(String())),
+    Column("environment", LowCardinality(Nullable(String()))),
+    Column("release", LowCardinality(Nullable(String()))),
+    Column("dist", LowCardinality(Nullable(String()))),
+    Column("ip_address_v4", Nullable(IPv4())),
+    Column("ip_address_v6", Nullable(IPv6())),
+    Column("user", WithDefault(String(), "''",)),
+    Column("user_hash", Materialized(UInt(64), "cityHash64(user)")),
+    Column("user_id", Nullable(String())),
+    Column("user_name", Nullable(String())),
+    Column("user_email", Nullable(String())),
+    Column("sdk_name", WithDefault(LowCardinality(String()), "''")),
+    Column("sdk_version", WithDefault(LowCardinality(String()), "''")),
+    Column("tags", Nested([("key", String()), ("value", String())])),
+    Column("_tags_flattened", String()),
+    Column("contexts", Nested([("key", String()), ("value", String())])),
+    Column("_contexts_flattened", String()),
+    Column("partition", UInt(16)),
+    Column("offset", UInt(64)),
+    Column("message_timestamp", DateTime()),
+    Column("retention_days", UInt(16)),
+    Column("deleted", UInt(8)),
+]
+
+
+class Migration(migration.MultiStepMigration):
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.TRANSACTIONS,
+                table_name="transactions_local",
+                columns=columns,
+                engine=table_engines.ReplacingMergeTree(
+                    version_column="deleted",
+                    order_by="(project_id, toStartOfDay(finish_ts), transaction_name, cityHash64(span_id))",
+                    partition_by="(retention_days, toMonday(finish_ts))",
+                    sample_by="cityHash64(span_id)",
+                    ttl="finish_ts + toIntervalDay(retention_days)",
+                    settings={"index_granularity": "8192"},
+                ),
+            )
+        ]
+
+    def backwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.TRANSACTIONS, table_name="transactions_local",
+            )
+        ]
+
+    def forwards_dist(self) -> Sequence[operations.Operation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.TRANSACTIONS,
+                table_name="transactions_dist",
+                columns=columns,
```

Comment:
The distributed version is not exactly the same as the local table because of materialization. You do not need to materialize columns in the distributed version (they would not work if you try to do so)

```
CREATE TABLE default.transactions_dist (
    `project_id` UInt64, 
    `event_id` UUID, 
    `trace_id` UUID, 
    `span_id` UInt64, 
    `transaction_name` LowCardinality(String), 
    `transaction_hash` UInt64, 
    `transaction_op` LowCardinality(String), 
    `transaction_status` UInt8, 
    `start_ts` DateTime, 
    `start_ms` UInt16, 
    `finish_ts` DateTime, 
    `finish_ms` UInt16, 
    `duration` UInt32, 
    `platform` LowCardinality(String), 
    `environment` LowCardinality(Nullable(String)), 
    `release` LowCardinality(Nullable(String)), 
    `dist` LowCardinality(Nullable(String)), 
    `sdk_name` LowCardinality(String), 
    `sdk_version` LowCardinality(String), 
    `ip_address_v4` Nullable(IPv4), 
    `ip_address_v6` Nullable(IPv6), 
    `user` String DEFAULT '', 
    `user_id` Nullable(String), 
    `user_name` Nullable(String), 
    `user_email` Nullable(String), 
    `tags.key` Array(String), 
    `tags.value` Array(String), 
    `_tags_flattened` String, 
    `contexts.key` Array(String), 
    `contexts.value` Array(String), 
    `_contexts_flattened` String, 
    `partition` UInt16, 
    `offset` UInt64, 
    `message_timestamp` DateTime, 
    `retention_days` UInt16, 
    `deleted` UInt8
) ENGINE = Distributed(snuba_transactions, default, transactions_local, cityHash64(span_id)) 
```

Reply:
Where's this statement pulled from? Is it the create table output in prod? I've noticed the materialize clause is always present there which is why I thought it was valid. (though i think it's not needed and figures it out from the remote table).

Reply:
I took it from prod, and I believe the dist table won't work if we have the materialize statement in both. James may remember the details.

Reply:
Ok. I updated this to strip out any materialization and use the underlying column type since that's what we have in prod.